### PR TITLE
qt: add adwaita platform theme

### PR DESCRIPTION
**Pull request by tschai-yim**
### Description

This PR adds `adwaita` as an option to `qt.platformTheme` fixing #4702 and also https://github.com/NixOS/nixpkgs/issues/295345 (the `gnome` option breaks some Qt6 tray applications). It also extends the configuraton interface to match `qt.style` which has a `name` and `package` sub-option for edge cases.

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

#### Maintainer CC

@rycee @thiagokokada

### CODE REVIEW

**Comment by thiagokokada** [comment]
> It also extends the configuraton interface to match qt.style which has a name and package sub-option for edge cases.

One reason I did not do this before is because those are all breaking changes, and there is no way to do this in a way that we could inform the users (e.g.: `mkRenamedOption` wouldn't work here).

So I am not sure about those changes. This will probably generate difficult to debug errors for the users.

**Comment by tschai-yim** [comment]
Added backwards compatibility with the following warning:
```shell
$ nix develop --ignore-environment .#qt-platform-theme-gnome
trace: warning: The option `qt.platformTheme` has been renamed to `qt.platformTheme.name`.
trace: warning: The value `gnome` for option `qt.platformTheme` is deprecated. Use `adwaita` instead.
bash: warning: setlocale: LC_NUMERIC: cannot change locale (de_CH.UTF-8): No such file or directory
bash: warning: setlocale: LC_TIME: cannot change locale (de_CH.UTF-8): No such file or directory
qt-platform-theme-gnome: OK
```
I think that should be enough for users to know what to do.

File: modules/misc/qt.nix [hunk 1]
```diff
@@ -55,56 +56,85 @@ in {
...
+            '';
+          };
+        };
+      in lib.mkOption {
         type = with lib.types;
-          nullOr (enum [ "gtk" "gtk3" "gnome" "lxqt" "qtct" "kde" ]);
+          nullOr
+          (either (enum [ "gtk" "gtk3" "gnome" "adwaita" "lxqt" "qtct" "kde" ])
+            (lib.types.submodule { options = newOption; }));
         default = null;
```
Comment (thiagokokada) [inline modules/misc/qt.nix:133]: Maybe we should set this option to `internal` so it doesn't show in docs?

**Review by thiagokokada — APPROVED** [review]
LGTM, thanks for the backwards compatibility changes.

File: modules/misc/qt.nix (hunk 1 above)
Comment (tschai-yim) [reply modules/misc/qt.nix:133]: No, I think it's better if users of the old option can still find it and then see the deprecation message.

Comment (thiagokokada) [reply modules/misc/qt.nix:133]: IMO, no because we don't want users to use those options. We want new users to see the new option only, and old users will receive the deprecation message.

Comment (tschai-yim) [reply modules/misc/qt.nix:133]: True, I'll set it to internal then. Though now `nix build .#docs-html` throws an error because [release-notes/rl-2105.md](https://github.com/nix-community/home-manager/blob/master/docs/release-notes/rl-2105.md?plain=1#L145) references the option. Because it is an old release note, we can probably remove the reference.

Another thing that I noticed is that the `.name` and `.package` submodule options don't get listed (even if `qt.platformTheme` isn't internal) in the reference documentation. The same happens with `services.syncthing.tray` which also has hidden submodules. Do you know of an option where this doesn't happen to see how we could solve this?

Comment (tschai-yim) [reply modules/misc/qt.nix:133]: Or we just accept it like that and keep `qt.platformTheme` not internal. I have a feeling the doc generator isn't capable of resolving submodules, especially because they're like normal [modules](https://discourse.nixos.org/t/override-submodule-options/12165). 

At least the correct option is listed in its description, and it says `or (submodule)` as type:
image
Not the most user-friendly, though.

Comment (thiagokokada) [reply modules/misc/qt.nix:133]: If this hides the submodule documentation themselves then I don't think setting it as internal is worth it. Sorry about the extra work.

Comment (tschai-yim) [reply modules/misc/qt.nix:133]: The submodules are hidden no matter the `internal` setting. I'll resolve this thread then.

File: modules/misc/qt.nix [hunk 2]
```diff
@@ -149,18 +179,37 @@ in {
...
+        (warnGnomeDeprecation "qt.platformTheme" cfg.platformTheme);
+      package = null;
+    } else if cfg.platformTheme == null then {
+      name = null;
+      package = null;
+    } else {
+      name =
+        warnGnomeDeprecation "cfg.platformTheme.name" cfg.platformTheme.name;
+      package = cfg.platformTheme.package;
+    };
```
Comment (rycee) [inline modules/misc/qt.nix:207]: Could the warnings here be expressed through the `warnings` option instead? Then it would be displayed consistently with other warnings and also be testable. See, e.g., https://github.com/nix-community/home-manager/blob/1c2acec99933f9835cc7ad47e35303de92d923a4/modules/programs/aerc.nix#L172-L182 and https://github.com/nix-community/home-manager/blob/1c2acec99933f9835cc7ad47e35303de92d923a4/tests/modules/programs/aerc/assertion.nix#L12-L20

**Comment by rycee** [comment]
Thanks! Merged to master now 🙂

**Comment by oxalica** [comment]
It seems my QT theme breaks after the update, and it's mixing the light theme and dark theme together. I guess this PR may be relevant. Either using the original `qt.platformTheme = "gnome"` or the new `qt.platformTheme.name = "adwaita";` has the same visual effect as screenshot below. Both dolphin from QT5 and QT6 suffer the visual issue. I'm not sure if my setting is the correct way to use Breeze-dark for QT apps.

The items are interleaving dark and light theme. (The grey box is just for redacted texts.)

image

My previously working configuration is [here](https://github.com/oxalica/nixos-config/blob/ab3c83836759087451e2068e960546634c00693f/home/modules/sway/default.nix#L62-L68).

**Comment by Birdy2014** [comment]
The README of QAdwaitaDecorations says that it is a "Qt decoration plugin", not a platform theme and that it should be applied by setting the QT_WAYLAND_DECORATION environment variable.

I don't know much about Qt, but it seems to me like it doesn't make much sense to try to set QAdwaitaDecorations as a platform theme.

**Comment by Ramblurr** [comment]
@lleheny0 did you ever figure this out?

**Comment by lukeleheny** [comment]
> @lleheny0 did you ever figure this out?

I did not unfortunately. Still using the same config that throws the warning, and it continues to work fine.
